### ISSUE TITLE
Trainer.predict should use padding token from tokenizer when possible

### ISSUE THREAD

**Issue by jacob-rosenthal** [huggingface/transformers#25379]
### System Info

- `transformers` version: 4.30.2
- Platform: macOS-13.0-arm64-arm-64bit
- Python version: 3.11.4
- Huggingface_hub version: 0.15.1
- Safetensors version: 0.3.1
- PyTorch version (GPU?): 2.0.0 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed

### Who can help?

@sgugger 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Trainer.predict()

### Expected behavior

As documented, `Trainer.predict()` uses a padding value of -100. However, if the Trainer is initialized with a tokenizer, it would be more consistent to use whatever the padding token is from the tokenizer. That way, the predictions can be decoded by the tokenizer more seamlessly.

I think this could be done by adding a check to `Trainer.evaluation_loop()` to set the padding index:

```python
pad_index = -100 if not self.tokenizer or not self.tokenizer.pad_token_id else self.tokenizer.pad_token_id
```
and then replacing the hardcoded -100 values in the code here with the `pad_index` from above: 

https://github.com/huggingface/transformers/blob/e42587f596181396e1c4b63660abf0c736b10dae/src/transformers/trainer.py#L3113-L3198

I would be happy to open a PR if you think this is a good idea. Thanks!

**Comment by sgugger** [comment]
No, that would be inconsistent with the index used by the losses functions for index that should be ignored.
